File: frontend/src/types/profiler.ts
/**
 * Profiler Types
 *
 * TypeScript interfaces for the Compound Profiler API responses.
 * All shapes match backend/app/schemas/profiler.py.
 */

// =============================================================================
// PFI (Property Forecast Index)
// =============================================================================

/**
 * PFI result with risk classification.
 * PFI = cLogP + number of aromatic rings. Low PFI → better developability.
 */
export interface PFIResult {
  pfi: number;
  clogp: number;
  aromatic_rings: number;
  risk: 'low' | 'moderate' | 'high';
}

// =============================================================================
// #Stars (QikProp-equivalent outlier count)
// =============================================================================

/**
 * A single property detail for #stars evaluation.
 * Shows whether a property falls within the acceptable range.
 */
export interface StarDetail {
  property: string;
  value: number;
  range_low: number;
  range_high: number;
  violated: boolean;
}

/**
 * #Stars result. Each violated property contributes one star.
 * Lower star count indicates a better overall property profile.
 */
export interface StarsResult {
  stars: number;
  details: StarDetail[];
}

// =============================================================================
// Abbott Bioavailability Score
// =============================================================================

/**
 * Abbott Bioavailability Score result.
 * Estimates probability of oral bioavailability using a 4-class model.
 */
export interface AbbottResult {
  abbott_score: number;
  probability_pct: number;
  tpsa: number;
  lipinski_violations: number;
}

// =============================================================================
// Consensus LogP
// =============================================================================

/**
 * Consensus LogP result combining multiple calculation methods.
 * Uses Wildman-Crippen and XLOGP3 to produce a consensus estimate.
 */
export interface ConsensusLogPResult {
  consensus_logp: number;
  wildman_crippen: number;
  xlogp3_approx: number;
  xlogp3_is_approximation: boolean;
}

// =============================================================================
// Skin Permeation
// =============================================================================

/**
 * Skin permeation result using the Potts-Guy model.
 * log Kp predicts dermal penetration rate.
 */
export interface SkinPermeationResult {
  log_kp: number;
  classification: 'low' | 'moderate' | 'high';
}

// =============================================================================
// 3D Shape Descriptors
// =============================================================================

/**
 * 3D shape descriptor result using PMI/NPR/PBF metrics.
 * Requires conformer generation — may fail for complex molecules.
 */
export interface Shape3DResult {
  pmi1: number;
  pmi2: number;
  pmi3: number;
  npr1: number;
  npr2: number;
  pbf: number;
  shape_class: 'rod' | 'disc' | 'sphere';
  '3d_conformer_failed': boolean;
}

// =============================================================================
// SA Score Comparison
// =============================================================================

/**
 * Detail for a single synthetic accessibility scoring method.
 */
export interface SAScoreDetail {
  score: number;
  scale: string;
  classification: 'easy' | 'moderate' | 'difficult';
  available: boolean;
  error?: string;
}

/**
 * SA Comparison result combining SA Score, SCScore, SYBA, and RAscore slot.
 * RAscore is not available (incompatible Python/TF version requirements).
 */
export interface SAComparisonResult {
  sa_score: SAScoreDetail;
  scscore: SAScoreDetail;
  syba: SAScoreDetail;
  rascore: { available: false; note: string };
}

// =============================================================================
// CNS MPO Score
// =============================================================================

/**
 * CNS MPO (Multi-Parameter Optimization) score result.
 * Uses Wager 2010 definition with 6 CNS-relevant properties.
 * Score out of max_score (typically 6).
 */
export interface CNSMPOResult {
  score: number;
  max_score: number;
  components: Record<string, number>;
}

// =============================================================================
// Custom MPO
// =============================================================================

/**
 * A single component in a custom MPO score computation.
 */
export interface CustomMPOComponent {
  property: string;
  raw_value: number;
  desirability: number;
  weight: number;
}

/**
 * Custom MPO score result.
 * Normalized score ranges from 0 to 1.
 */
export interface CustomMPOResult {
  score: number;
  max_score: number;
  normalized: number;
  components: CustomMPOComponent[];
}

// =============================================================================
// Ligand Efficiency
// =============================================================================

/**
 * Extended ligand efficiency metrics.
 * Requires an activity value (IC50/Ki/pIC50/pKd) to compute.
 */
export interface LEResult {
  pIC50: number;
  LE: number;
  LLE: number;
  LELP: number;
  BEI: number;
  SEI: number;
}

// =============================================================================
// MPO Property Configuration
// =============================================================================

/**
 * A single property configuration for custom MPO scoring.
 * Defines the desirability function shape and acceptable range.
 */
export interface MPOProperty {
  property: string;
  low: number;
  high: number;
  weight: number;
  shape: 'sigmoid' | 'ramp' | 'step';
}

// =============================================================================
// Full Profile Response
// =============================================================================

/**
 * Complete profiler response from /api/v1/profiler/full.
 * Contains results for all computed profiling metrics.
 */
export interface ProfileResponse {
  pfi: PFIResult;
  stars: StarsResult;
  abbott: AbbottResult;
  consensus_logp: ConsensusLogPResult;
  skin_permeation: SkinPermeationResult;
  sa_comparison: SAComparisonResult;
  cns_mpo: CNSMPOResult;
  druglikeness?: Record<string, unknown>;
}

// =============================================================================
// Identifier Resolution
// =============================================================================

/**
 * Response from /api/v1/resolve for identifier-to-SMILES conversion.
 */
export interface ResolveResponse {
  smiles: string;
  identifier: string;
  source: string;
}

File: backend/app/schemas/profiler.py
"""
Profiler Schemas

Pydantic v2 request/response models for the compound profiler endpoints.
Covers: PFI, #stars, Abbott bioavailability, consensus LogP, skin permeation,
3D shape descriptors, CNS MPO, custom MPO, and extended ligand efficiency.
"""

from typing import Any, Optional

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Request Models
# ---------------------------------------------------------------------------


class ProfileRequest(BaseModel):
    """Request for full compound profile computation."""

    smiles: str = Field(..., min_length=1, description="SMILES string to profile")


class LERequest(BaseModel):
    """Request for extended ligand efficiency computation."""

    smiles: str = Field(..., min_length=1, description="SMILES string")
    activity_value: float = Field(..., description="Activity measurement value")
    activity_type: str = Field(
        ...,
        pattern="^(IC50_nM|IC50_uM|Ki_nM|pIC50|pKd)$",
        description="Activity type: IC50_nM | IC50_uM | Ki_nM | pIC50 | pKd",
    )


class MPOProperty(BaseModel):
    """A single property entry in a custom MPO profile."""

    property: str = Field(..., description="Property name (MW, LogP, TPSA, HBD, etc.)")
    low: float = Field(..., description="Lower boundary of the favorable range")
    high: float = Field(..., description="Upper boundary of the favorable range")
    weight: float = Field(default=1.0, ge=0, description="Relative importance weight")
    shape: str = Field(
        default="sigmoid",
        pattern="^(sigmoid|ramp|step)$",
        description="Desirability curve shape: sigmoid | ramp | step",
    )


class MPORequest(BaseModel):
    """Request for custom MPO score computation."""

    smiles: str = Field(..., min_length=1, description="SMILES string")
    profile: list[MPOProperty] = Field(
        ...,
        min_length=1,
        description="List of MPO property desirability definitions",
    )


class Shape3DRequest(BaseModel):
    """Request for 3D shape descriptor computation."""

    smiles: str = Field(..., min_length=1, description="SMILES string")


class SAComparisonRequest(BaseModel):
    """Request for SA Score / SCScore / SYBA comparison."""

    smiles: str = Field(..., min_length=1, description="SMILES string")


# ---------------------------------------------------------------------------
# Response Models
# ---------------------------------------------------------------------------


class PFIResult(BaseModel):
    """Property Forecast Index result."""

    pfi: float = Field(description="PFI score (cLogP + aromatic rings)")
    clogp: float = Field(description="Wildman-Crippen cLogP")
    aromatic_rings: int = Field(description="Number of aromatic rings")
    risk: str = Field(description="Risk classification: low | moderate | high")


class StarsPropertyDetail(BaseModel):
    """Per-property detail for #stars."""

    property: str = Field(description="Property name")
    value: float = Field(description="Computed property value")
    range_low: float = Field(description="95th-percentile lower bound")
    range_high: float = Field(description="95th-percentile upper bound")
    violated: bool = Field(description="True if value is outside the range")


class StarsResult(BaseModel):
    """#Stars outlier count result."""

    stars: int = Field(ge=0, description="Number of 95th-percentile violations")
    details: list[StarsPropertyDetail] = Field(
        description="Per-property detail for all 8 properties"
    )


class AbbottResult(BaseModel):
    """Abbott Bioavailability Score result."""

    abbott_score: float = Field(description="Bioavailability probability (0.11/0.17/0.56/0.85)")
    probability_pct: int = Field(description="Bioavailability probability as integer percent")
    tpsa: float = Field(description="Topological polar surface area used in calculation")
    lipinski_violations: int = Field(description="Number of Lipinski Ro5 violations")


class ConsensusLogPResult(BaseModel):
    """Consensus LogP result."""

    consensus_logp: float = Field(description="Average of Wildman-Crippen and XLOGP3 approx")
    wildman_crippen: float = Field(description="Wildman-Crippen cLogP (RDKit)")
    xlogp3_approx: float = Field(
        description="XLOGP3 approximation (0.92 * Wildman-Crippen + 0.12)"
    )
    xlogp3_is_approximation: bool = Field(
        description="Always True — XLOGP3 value is approximated, not computed by XLogP3 itself"
    )


class SkinPermeationResult(BaseModel):
    """Skin permeation coefficient result."""

    log_kp: float = Field(description="Skin permeation coefficient (log Kp, cm/h)")
    classification: str = Field(description="Permeation class: low | moderate | high")


class Shape3DResult(BaseModel):
    """3D shape descriptor result."""

    pmi1: float = Field(description="Principal Moment of Inertia 1 (smallest)")
    pmi2: float = Field(description="Principal Moment of Inertia 2")
    pmi3: float = Field(description="Principal Moment of Inertia 3 (largest)")
    npr1: float = Field(ge=0, le=1, description="Normalized PMI ratio 1 (rod axis)")
    npr2: float = Field(ge=0, le=1, description="Normalized PMI ratio 2 (disc axis)")
    pbf: float = Field(description="Plane of Best Fit (PBF)")
    shape_class: str = Field(description="Shape classification: sphere | disc | rod")


class CNSMPOResult(BaseModel):
    """CNS MPO score result (Wager 2010, 4-component RDKit version)."""

    score: float = Field(ge=0, le=4, description="CNS MPO score (0-4 for 4-component version)")
    max_score: float = Field(description="Maximum possible score (4.0 for 4-component)")
    components: dict[str, float] = Field(
        description="Per-component desirability values: clogp, tpsa, mw, hbd"
    )


class MPOComponentDetail(BaseModel):
    """Per-property detail for custom MPO."""

    property: str = Field(description="Property name")
    value: float = Field(description="Computed property value")
    desirability: float = Field(ge=0, le=1, description="Desirability (0-1)")
    weight: float = Field(description="Property weight")
    contribution: float = Field(description="Weighted contribution")


class CustomMPOResult(BaseModel):
    """Custom MPO score result."""

    score: float = Field(description="Weighted sum of desirabilities")
    max_score: float = Field(description="Sum of all weights (maximum possible score)")
    normalized: float = Field(ge=0, le=1, description="score / max_score (0-1)")
    components: list[MPOComponentDetail] = Field(description="Per-property breakdown")


class LEResult(BaseModel):
    """Extended ligand efficiency result."""

    pIC50: float = Field(description="Converted pIC50 value")  # noqa: N815
    LE: float = Field(description="Ligand Efficiency (1.4 * pIC50 / HA)")
    LLE: float = Field(description="Lipophilic Ligand Efficiency (pIC50 - LogP)")
    LELP: float = Field(description="Ligand Efficiency Lipophilicity Price (LogP / LE)")
    BEI: float = Field(description="Binding Efficiency Index (pIC50 * 1000 / MW)")
    SEI: float = Field(description="Surface Efficiency Index (pIC50 * 100 / TPSA)")


class ProfileResponse(BaseModel):
    """Composite compound profile response (full endpoint)."""

    pfi: PFIResult = Field(description="PFI score and risk classification")
    stars: StarsResult = Field(description="#Stars outlier count with property details")
    abbott: AbbottResult = Field(description="Abbott Bioavailability Score")
    consensus_logp: ConsensusLogPResult = Field(description="Consensus LogP")
    skin_permeation: SkinPermeationResult = Field(description="Skin permeation (log Kp)")
    druglikeness: Optional[dict[str, Any]] = Field(
        None, description="Drug-likeness rules summary (Lipinski, QED, Veber)"
    )
